### DISCUSSION TITLE
gVisor based tun2socks

### DISCUSSION THREAD

**Discussion by klzgrad** [SagerNet/SagerNet#142] [Ideas]
Are you aware of this https://github.com/xjasonlyu/tun2socks? Based on gVisor created by Google.

I think it is interesting technology and should be much more modern than lwip. Integrating it would take time but it'd be a great learning experience.

**Comment by nekohasekai** [comment]
It seems that it lacks some parameters like `interface private address` and `dnsgw`. And it has a built-in go-shadowsocks2, I don't know what's the use of this.

**Reply by klzgrad** [reply]
gVisor's TCP/IP stack is interesting. This tun2socks fork is not interesting. I'm only saying to replace lwip with this TCP/IP stack.

**Comment by rurirei** [comment]
i'm interested in that and i do not know the all codes (how to implement a tun2socks with Netstack, and a proxy dialer with V2ray). i'm pleased to see the possible relative work.

**Reply by nekohasekai** [reply]
V2Ray v5 may support tun mode so I just waited :)

**Reply by rurirei** [reply]
what i said is tun2socks/proxy/v2ray.go (i'm not interested in socks inbound).

you cannot make a "tun" device on Android, what you depends on is android VPNService (if no root).

**Reply by nekohasekai** [reply]
Android VPN Service will create a tun and then we get its fd, this is how tun2socks works

**Comment by nekohasekai** [comment]
I tried gVisor but gave up, because it's not compatible with the tun created by the android framework (the root cause is my lack of skill, I'm just an amateur, so).

I'm now implementing a transparent proxy-like forwarder (inspired by NetBare, the open source GPL library of a famous android packet capturer).

It works by parsing a TCP/IP packet, then rewriting it so that it is sent to a local TCP forwarding server, which can connect to the real address with the recorded connection information; for UDP, it is a direct proxy (I use Netty to implement them). 

Since rewriting packets is fast (I'll provide a multi-threaded toggle, but I don't know if it's necessary), and the forwarding server is implemented using the non-blocking IO server framework, I think it might perform better than tun2socks under heavy traffic.

There are some additional features that can be easily implemented based on this, such as a firewall or real-time application traffic speed display.

**Reply by rurirei** [reply]
> I tried gVisor but gave up, because it's not compatible with the tun created by the android framework

of course you need Root Access for creating a tun device, that's needed by netstack (or i did failed to try on netstack.tcpip.tcp/udp.forwarder())

> inspired by NetBare, the open source GPL library of a famous android packet capturer

if that needs Root Access, i may interested on the approach of it?

**Reply by nekohasekai** [reply]
> of course you need Root Access

https://github.com/SagerNet/SagerNet/discussions/142#discussioncomment-910053

**Reply by nekohasekai** [reply]
> if that needs Root Access, i may interested on the approach of it?

?

**Comment by rurirei** [comment]
but... i made this done with some bugs lol: ~~with open source later, maybe?~~

**Reply by nekohasekai** [reply]
It's here: https://github.com/SagerNet/LibSagerNetCore/tree/main/gvisor

**Reply by rurirei** [reply]
so good 👋🏻
